content/post/079-async-await-refactor.md

```diff
@@ -0,0 +1,122 @@
+
+## 2. Streaming with async iterables
+
+Our streaming APIs will use [async iterables](https://javascript.info/async-iterators-generators). Thats significant for a couple of reasons. Firstly, since we've been talking about bundle sizes it's worth mentioning that using a streaming implementation that is actually a language feature means that we don't need to _bundle_ any libraries/modules to provide a streaming implementation! Obviously that also means that we don't need to `npm install` them either.
+
+Secondly, **async iterables make streaming APIs approachable**. There's a BIG hurdle involved with streaming APIs that is lowered significantly by using async iterables instead of Node.js or pull streams. The hurdle is the understanding of the concepts, naming and usage that surrounds them that developers new to JS or new to streaming have to overcome.
+
+There's a big conflict of interests in exposing streaming APIs and exposing an API that is easy to use and understand for _all_ developers. When you boil down the understanding of reading from a stream to a `for` loop (you read an async iterable using [`for await...of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of) loops) then developers can instantly draw parallels to one of the first things they learnt in programming. This is simply not the case if your API returns a Node.js `ReadableStream` for example.
+
+## 3. Better debugging
+
+One of the best things about debugging in an `async`/`await` code base is that **stack traces are much better**. With callbacks you typically lose the stack trace across an async boundary because with callbacks, the call stack starts at the callback function.
+
+This can make debugging super difficult because you can't get a good idea of the execution path of the program prior to the async boundary. The situation is improved by using promises but by using `async`/`await` it's _even better_:
+
+> `await X()` suspends execution of the current function, while `promise.then(X)` continues execution of the current function after adding the X call to the callback chain. In the context of stack traces, this difference is pretty significant.
+> https://mathiasbynens.be/notes/async-stack-traces
+
+The upshot is that stack traces are more informative and span async boundaries giving developers more power to debug code.
+
+The second best thing about debugging in an `async`/`await` code base is that synchronous and asynchronous errors are handled by the same construct: `try`/`catch`.
+
+Asynchronous errors when using callbacks must be handled by checking for an error parameter passed to the callback function using an `if` statement. You must also either _trust_ that a call to a function that takes a callback does not throw synchronously (this is what most people do) or put a `try`/`catch` around it as well.
```

Comment:
"trust that a call doesn't throw synchronously" -- is that a correct saying in js-land or is "trust that a call doesn't throw a synchronous error" more grammatically correct?

Reply:
(I see a lot of this "throw" below - so if normal maybe explain what it is the first time so I don't get lost with this new term?)

Reply:
Reading through - this is the section that really lost me. there's a lot of detail between each comparison and it's hard to know where to skip to to get out of the debugging details. Maybe either try cutting back or combining into one technically dense paragraph so non-js readers can easily skip to the next section if this isn't relevant to them?

Reply:
I think "synchronously" can just be removed from that sentence.

I'll try to fix up...

Programming languages typically have some sort of concept of throwing or raising an error/exception that can be caught. The golang equivalent is panic/recover.

Reply:
Ok, I've expanded this section considerably with examples and better explanations. Hopefully it's better now 🤞